[![Build Status](https://travis-ci.org/iqbal-lab-org/gramtools.svg?branch=master)](https://travis-ci.org/iqbal-lab-org/gramtools)

# gramtools
**TL;DR** Genome inference using prior information encoded as a reference graph.

Gramtools builds a directed acyclical graph (DAG) of genetic variation from a population of genomes. Given sequence data from a haploid individual, alleles on the graph are annotated with coverage information. A personalised reference genome for the sample being studied can then be inferred from the annotated graph.

## Install
```
pip3 install -vvv --process-dependency-links wheel git+https://github.com/iqbal-lab-org/gramtools
```
<br>

If `sudo` is unavailable, we recommend using a Python virtual enviroment:
```
python3 -m venv gram_ve && source gram_ve/bin/activate
```
Note: installation fails with `virtualenv`.

## Usage
Gramtools currently consists of four commands. These commands are documented in the wiki (see links below). In dependency order, they are:
1) [build](https://github.com/iqbal-lab-org/gramtools/wiki/Commands%3A-build) - given a VCF and reference, construct the graph

2) [quasimap](https://github.com/iqbal-lab-org/gramtools/wiki/Commands%3A-quasimap) - map reads to a graph generated in `build`

3) [infer](https://github.com/iqbal-lab-org/gramtools/wiki/Commands%3A-infer) - infer a maximum likelihood personalised reference genome from the mapping coverage generated in `quasimap`

4) [discover](https://github.com/iqbal-lab-org/gramtools/wiki/Commands%3A-discover) - find new variation against the personalised reference genome using 1+ standard variant callers (currently: cortex)

Examples, documentation, and planned future enhancements can be found in the [wiki](https://github.com/iqbal-lab-org/gramtools/wiki).

```
Gramtools

Usage:
  gramtools build --gram-dir --vcf --reference --kmer-size [--all-kmers]
  gramtools quasimap --gram-dir --reads --run-dir
  gramtools infer --run-dir [--infer-mode]
  gramtools discover --run-dir

  gramtools (-h | --help)
  gramtools --version

Options:
  --gram-dir 	Gramtools directory containing outputs of `build` 
  --kmer-size 	Kmer size at which to build the graph (used for seeding reads in `quasimap`)	

  --run-dir 	Run directory; stores outputs of `quasimap`, `infer`, `discover`
  --reads 	1+ reads file, in (fasta/fastq/sam/bam/cram) format

  -h --help             Show this screen
  --version             Show version

Subcommands:
  build         Construct the graph and supporting data structures
  quasimap      Map reads to graph
  infer         Infer a maximum likelihood personalised reference genome
  discover	Discover new variation against personalised reference genome
```

## Reference documentation

A [doxygen](http://doxygen.nl/) formatted documentation can be generated by running 
```doxygen doc/Doxyfile.in```
from inside the gramtools directory.

The documentation gets generated in doc/html/index.html and provides a useful reference for all files, classes, functions and data structures in gramtools.

## License

MIT
